### PR TITLE
redis_proxy: add support for external authentication

### LINKED ISSUE #35673: redis_proxy: add support for external authentication
## Feature overview

Redis proxy users may want to create advanced authentication methods.
For example, the official [Azure SDK extension for Redis](https://github.com/Azure/Microsoft.Azure.StackExchangeRedis) allows to authenticate to a Redis server using Microsoft Entra ID token-based authentication, by passing a token in the password argument of the `AUTH` command periodically, based on token expiration.
This PR introduces a way to support external authentication via a gRPC service with additional support for expiry of such authentication (e.g. for token-based authentication).
This way we keep it extensible for **any** advanced authentication methods users might want to develop.

### FAQ: Why not use the _ext_authz_ filter?
The cost/latency impact by using the _ext_authz_ filter is much bigger than the proposed design. That's because instead of being called on every request, the current design only calls the external dependency on **AUTH** commands. Not only that, but also we would have to decode the Redis protocol twice, if we used a separate filter.

### PR DESCRIPTION
**Pull request by dbarbosapn**
resolves #35673 

## PR overview

Redis proxy users may want to create advanced authentication methods.
For example, the official [Azure SDK extension for Redis](https://github.com/Azure/Microsoft.Azure.StackExchangeRedis) allows to authenticate to a Redis server using Microsoft Entra ID token-based authentication, by passing a token in the password argument of the `AUTH` command periodically, based on token expiration.
This PR introduces a way to support external authentication via a gRPC service with additional support for expiry of such authentication (e.g. for token-based authentication).
This way we keep it extensible for **any** advanced authentication methods users might want to develop.

### The reviewer may ask: Why not use the _ext_authz_ filter?
The cost/latency impact by using the _ext_authz_ filter is much bigger than the proposed design. That's because instead of being called on every request, the current design only calls the external dependency on **AUTH** commands. Not only that, but also we would have to decode the Redis protocol twice, if we used a separate filter.

---

Risk Level: Medium (small optional feature added to existing filter)

Testing:  ✅ 
- Unit Tests
- Integration Tests
- Manual Testing
image
- Also, we are already using the signed _-dev_ build on a test AKS cluster

Docs Changes: ✅ 

- Proto docs
image

- Updated authentication-related information on the Redis protocol page.

Release Notes: ✅

### CODE REVIEW

**Comment by dbarbosapn** [comment]
@mattklein123 all tests are passing locally, but on the pipeline:
image

I cannot reproduce locally. I believe it's something related to the simulated time system but I can't find what.

Any help is appreciated!

**Comment by adisuissa** [comment]
Assigning Matt as redis_proxy codeowner.
/assign @mattklein123

**Comment by dbarbosapn** [comment]
@mattklein123 gentle ping 😄 Could I get your review?

**Comment by dbarbosapn** [comment]
> Thanks this looks very nice. Please add an actual integration test for this feature given the complexities that real networking adds. https://github.com/envoyproxy/envoy/blob/main/test/extensions/filters/network/redis_proxy/redis_proxy_integration_test.cc
> 
> It will be not trivial but there are lots of examples of doing fake upstreams like this in integration tests. Thank you.
> 
> /wait

Thank you very much for the review, @mattklein123 !
Will work on adding the integration tests. Should have it by tomorrow or Tuesday max.

Thanks!

**Comment by dbarbosapn** [comment]
Hi @mattklein123!
Added the integration testing. Pretty cool system you guys created for fake upstreams 😄 

Let me know if all looks good now.

Thank you!

**Review by mattklein123 — APPROVED** [review]
Thanks!

File: source/extensions/filters/network/redis_proxy/proxy_filter.h [hunk 1]
```diff
@@ -1,5 +1,7 @@
 #pragma once
 
+#include <bits/types/time_t.h>
```
Comment (zirain) [inline source/extensions/filters/network/redis_proxy/proxy_filter.h:3]: this seems break build on macos m1
```
./source/extensions/filters/network/redis_proxy/proxy_filter.h:3:10: fatal error: 'bits/types/time_t.h' file not found
```

Comment (dbarbosapn) [reply source/extensions/filters/network/redis_proxy/proxy_filter.h:3]: Will take a look. Will be hard to repro without a mac, and I but I'm guessing this is not the right import so I'll check what should be the alternative

Comment (zirain) [reply source/extensions/filters/network/redis_proxy/proxy_filter.h:3]: create a CI for this check: https://github.com/envoyproxy/envoy/pull/35933

**Pull request merged**